oak_idl/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+    fn from(e: TransportError) -> Self {
+        ClientError::TransportError(e)
+    }
+}
+
+#[derive(Debug)]
+pub enum TransportError {
+    InvalidMessage,
+    InvalidMethodId,
+}
+
+/// Unique identifier of a method within a service.
+type MethodId = u32;
+
+/// Header of a transactional message.
+///
+/// See <https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format#transactional-messages>
+#[derive(Serialize, Deserialize)]
+pub struct Header {
+    pub transaction_id: u32,
+    pub method_id: MethodId,
+}
+
+/// A transactional message.
+///
+/// The body field contains the serialized payload of the request or response.
+///
+/// Note that the serialization of the body does not have to match the serialization of the Message
+/// itself. For instance, the Message may be serialized via bincode, but the paload may be
+/// serialized via protobuf.
+#[derive(Serialize, Deserialize)]
+pub struct Message {
+    pub header: Header,
+    pub body: Vec<u8>,
+}
+
+/// A message-oriented transport that allows performing invocations.
+pub trait Transport {
+    fn invoke(&self, request_message: &Message) -> Result<Message, TransportError>;
+}
+
+/// A wrapper for a message-oriented channel handle, which implements the [`Transport`] trait.
+struct Handle(u32);
+
+impl Transport for Handle {
+    fn invoke(&self, request_message: &Message) -> Result<Message, TransportError> {
+        // The serialization of the outer message is fixed by the transport, since the header needs
+        // to be deserializable by the other end of the transport, but the body of the
+        // request / response is serialized at a higher level, and in potentially different format,
+        // since that's handled by the application.
+        let request_bytes =
+            bincode::serialize(&request_message).map_err(|_| TransportError::InvalidMessage)?;
+        let response_bytes = invoke(self.0, &request_bytes)?;
+        let response_message =
+            bincode::deserialize(&response_bytes).map_err(|_| TransportError::InvalidMessage)?;
+        Ok(response_message)
+    }
+}
+
+/// A low-level message-oriented transport over channels. This is meant to emulate the low-level
+/// Wasm ABI in Oak, and will eventually be replaced by that.
+fn invoke(_channel_handle: u32, _request: &[u8]) -> Result<Vec<u8>, TransportError> {
+    unimplemented!()
+}
+
+/// This macro generates the following objects (assuming it is invoked with the first argument set
+/// to `Name`):
+///
+/// - a struct named `NameClient`, exposing a method for each method defined in the macro.
+/// - a struct named `NameServer`, which implements the [`Transport`] trait, dispatching each
+///   request to the appropriate method on the underlying service implementation.
+/// - a trait named `Name`, with a method for each method defined in the macro, and an additional
+///   default method named `serve` which returns an instance of `NameServer`; the developer of a
```

Comment:
So `serve` would correspond to `invoke`, e.g., in the lookup extension here: https://github.com/project-oak/oak/blob/a978dc57fec21d62bcbe0d4f59e0d97245ef2e96/oak_functions/loader/src/lookup.rs#L83